*: Move ShouldDelegateList into storage layer and make it exportable

MadhavJivrajani opened the pull request:
#### What type of PR is this?

/kind cleanup
/priority backlog
#### What this PR does / why we need it:

storage: ShouldDelegateList is now a public function which tells if the LIST request will be served from the watch cache or etcd.

Commit 1:
Previously, shouldDelegateList was an unexported helper function
whose logic was used in list work estimation and integration tests.
This needed any change in its logic to be made individually to
areas which replicated it.

This commit makes the function exportable and makes it a utility
function defined by the storage package hence making it re-usable
by other areas of the codebase.

One of the reasons for moving it into the storage package was also to
avoid import cycles between the cacher package and the flowcontrol package.

Commit 2:
This commit re-uses the ShouldDelegateList utility function from the storage package.

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

/assign @wojtek-t

MadhavJivrajani committed:
storage: Move ShouldDelegateList into storage layer and make it exportable

Previously, shouldDelegateList was an unexported helper function
whose logic was used in list work estimation and integration tests.
This needed any change in its logic to be made individually to
areas which replicated it.

This commit makes the function exportable and makes it a utility
function defined by the storage package hence making it re-usable
by other areas of the codebase.

One of the reasons for moving it into the storage package was also to
avoid import cycles between the cacher package and the flowcontrol package.

Signed-off-by: Madhav Jivrajani <madhav.jiv@gmail.com>

MadhavJivrajani committed:
*: Re-use ShouldDelegateList utility function

This commit re-uses the ShouldDelegateList utility
function from the storage package.

Signed-off-by: Madhav Jivrajani <madhav.jiv@gmail.com>

MadhavJivrajani commented:
/sig api-machinery

MadhavJivrajani commented:
/assign @MikeSpreitzer 
in case you have thoughts.

wojtek-t commented:
Updated release notes - it's not something our users are interested in.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/request/list_work_estimator.go:
```diff
@@ -57,7 +57,14 @@ func (e *listWorkEstimator) estimate(r *http.Request, flowSchemaName, priorityLe
 		return WorkEstimate{InitialSeats: maximumSeats}
 	}
-	isListFromCache := !shouldListFromStorage(query, &listOptions)
+	isListFromCache := !storage.ShouldDelegateList(storage.ListOptions{
```
This is potentially risky change.
If at some point we would decide to use other fields of ListOptions in the function (like selectors), those aren't set here. Also - we didn't have to create storage.ListOptions here, which adds some cost.

So I'm not 100% convinced we want this change.

MadhavJivrajani replied:
@wojtek-t 

> If at some point we would decide to use other fields of ListOptions in the function (like selectors), those aren't set here

That's a fair point. How about having a conversion function from `metav1.ListOptions` -> `storage.ListOptions` that is used in these places?

> we didn't have to create storage.ListOptions here, which adds some cost

As far as I understand, we wouldn't get any additional allocs. Everything would happen through the stack. Would a benchmark help?

Trying to think of ways to move forward because I do think that keeping all sources in sync with the logic in that function would get difficult, I am currently aware of 2 places (flow control and integration test), its possible that other places will come up.

wojtek-t replied:
Conversion func would help, although I would like it to be constructed in a way that it could also be used in:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L356

> Would a benchmark help?

Yeah - some small benchmark would be useful here.

leilajal commented:
/assign @wojtek-t

MadhavJivrajani committed:
*: Add conversion func for metav1.ListOptions -> storage.ListOptions

Introduce a conversion func to convert between metav1.ListOptions
and storage.ListOptions in order to ensure that we have all fields
needed when storage.ListOptions is used in ShouldDelegateList.

Signed-off-by: Madhav Jivrajani <madhav.jiv@gmail.com>

MadhavJivrajani commented:
@wojtek-t I've added a conversion function. PTAL

Also I'm not entirely sure what you meant here:
> Conversion func would help, although I would like it to be constructed in a way that it could also be used in:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L356

Could you please elaborate in case I misunderstood?

---

As per the benchmark, I wrote and ran the following benchmark:
```go
func BenchmarkShouldDelegateList(b *testing.B) {
	listOptions := metav1.ListOptions{
		ResourceVersion:      "12345",
		ResourceVersionMatch: metav1.ResourceVersionMatchNotOlderThan,
		Limit:                45,
		Continue:             strings.Repeat("a", 64),
	}
	b.ResetTimer()
	b.ReportAllocs()

	for i := 0; i < b.N; i++ {
                // Or run with the previous vanilla function that was deleted.
		storage.ShouldDelegateList(storage.ConvertToListOptions(listOptions))
	}
}
```

The benchmark was run for a period of 5s, 10 times over and then the results passed through `benchstat`:

```
❯ benchstat vanilla.txt conversion.txt
name                   old time/op    new time/op    delta
ShouldDelegateList-16    15.9ns ± 1%    28.1ns ± 1%  +76.35%  (p=0.000 n=9+9)

name                   old alloc/op   new alloc/op   delta
ShouldDelegateList-16     0.00B          0.00B          ~     (all equal)

name                   old allocs/op  new allocs/op  delta
ShouldDelegateList-16      0.00           0.00          ~     (all equal)
```
There is some cpu cost added (as expected), although I don't think this is anything significant considering the ns scale its in.
Lmk what you think, thanks!

wojtek-t commented:
> Could you please elaborate in case I misunderstood?

Sorry. What I meant is that I would like this conversion function to also be used in the place I pointed at.
In such case, any problem with that function should surface pretty quickly, because it's *the* place that does that.

Regarding benchmark - can you please check that in to this PR?

leilajal commented:
/triage accepted

ncdc commented:
/uncc

MadhavJivrajani commented:
/reopen
